File: java/498_DiagonalTraverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return new int[0]; 
        int r = matrix.length, c = matrix[0].length;
        int[] diagonals = new int[r * c];
        
        int i = 0, j = 0;
        boolean up = true;
        for (int idx = 0; idx < r * c; ++idx) {
            diagonals[idx] = matrix[i][j];
            if (up) {
                if (i == 0 || j == c - 1) {
                    if (j == c - 1)
                        i++;
                    else
                        j++;
                    up = false;
                } else {
                    i--;
                    j++;
                }
            } else {
                if (i == r - 1 || j == 0) {
                    if (i == r - 1)
                        j++;
                    else
                        i++;
                    up = true;
                } else {
                    i++;
                    j--;
                }
            }
        }
        return diagonals;
    }
}

File: swift/498_DiagonalTraverse.swift
class Solution {
    func findDiagonalOrder(_ matrix: [[Int]]) -> [Int] {
        var diagonals = [Int]()
        if matrix.isEmpty || matrix[0].isEmpty {
            return diagonals
        }
        
        var N = matrix.count, M = matrix[0].count
        var up = true
        var i = 0, j = 0
        for idx in 0..<N * M {
            diagonals.append(matrix[i][j])
            if up {
                if j == M - 1 {
                    i += 1
                    up = false
                } else if i == 0 {
                    j += 1
                    up = false
                } else {
                    i -= 1
                    j += 1
                }
            } else {
                if i == N - 1 {
                    j += 1
                    up = true
                } else if j == 0 {
                    i += 1
                    up = true
                } else {
                    i += 1
                    j -= 1
                }
            }
        }
        return diagonals
    }
}
